src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1045,97 +1075,207 @@ private static void ParseStatusLine(ReadOnlySpan<byte> line, HttpResponseMessage
+                    }
+
+                    int crIdx = lfIdx - 1;
+                    int crOrLfIdx = (uint)crIdx < (uint)valueIter.Length && valueIter[crIdx] == '\r'
+                        ? crIdx
+                        : lfIdx;
+
+                    int spIdx = lfIdx + 1;
+                    if ((uint)spIdx >= (uint)valueIter.Length)
+                    {
+                        return (finished: false, bytesConsumed: originalBufferLength - buffer.Length);
+                    }
+
+                    if (valueIter[spIdx] is not (byte)'\t' and not (byte)' ')
+                    {
+                        // Found the end of the header value.
+
+                        if (response is not null)
+                        {
+                            ReadOnlySpan<byte> headerName = buffer.Slice(0, valueStartIdx - 1);
+                            ReadOnlySpan<byte> headerValue = buffer.Slice(valueStartIdx, buffer.Length - valueIter.Length + crOrLfIdx - valueStartIdx);
+                            AddResponseHeader(headerName, headerValue, response, isFromTrailer);
+                        }
+
+                        buffer = buffer.Slice(buffer.Length - valueIter.Length + spIdx);
+                        break;
+                    }
+
+                    // Found an obs-fold (CRLFHT/CRLFSP).
+                    // Replace the CRLF with SPSP and keep looking for the final newline.
+                    valueIter[crOrLfIdx] = (byte)' ';
+                    valueIter[lfIdx] = (byte)' ';
+
+                    valueIter = valueIter.Slice(spIdx + 1);
+                }
             }
 
-            if (!HeaderDescriptor.TryGet(line.Slice(0, pos), out HeaderDescriptor descriptor))
+            static void ThrowForInvalidHeaderLine(ReadOnlySpan<byte> buffer, int newLineIndex) =>
+                throw new HttpRequestException(SR.Format(SR.net_http_invalid_response_header_line, Encoding.ASCII.GetString(buffer.Slice(0, newLineIndex))));
+        }
+
+        private void AddResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value, HttpResponseMessage response, bool isFromTrailer)
+        {
+            // Skip trailing whitespace and check for empty length.
+            while (true)
             {
-                // Invalid header name.
-                throw new HttpRequestException(SR.Format(SR.net_http_invalid_response_header_name, Encoding.ASCII.GetString(line.Slice(0, pos))));
+                int spIdx = name.Length - 1;
+
+                if ((uint)spIdx < (uint)name.Length)
+                {
+                    if (name[spIdx] != ' ')
+                    {
+                        // hot path
+                        break;
+                    }
+
+                    name = name.Slice(0, spIdx);
+                }
+                else
+                {
+                    ThrowForEmptyHeaderName();
+                }
             }
```

Comment:
This looks like a really slow way to trim ending whitespace.  Do we not expect any?

Reply:
(Also it's only looking for ' '... other whitespace doesn't matter?)

Reply:
The expected hot path is that there are no trailing spaces.

> (Also it's only looking for ' '... other whitespace doesn't matter?)

This matches the [existing behavior](https://github.com/dotnet/runtime/blob/2b07c8c46523b816b1eac5c8af092e52b20a69d2/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L1083-L1092).

According to the spec, the name shouldn't have trailing whitespace at all
[`field-line   = field-name ":" OWS field-value OWS`](https://www.rfc-editor.org/rfc/rfc9112#name-field-syntax)

We also trim the whitespace around the value, but we trim leading SP **and** HTAB, but only trailing SP.

I think it's fine that we're more accepting on the name, but it's odd that we differ for leading/trailing OWS on the value.